sdks/python/apache_beam/transforms/deduplicate.py

```diff
@@ -0,0 +1,108 @@
+from apache_beam.transforms.core import ParDo
+from apache_beam.transforms.timeutil import TimeDomain
+from apache_beam.typehints import trivial_inference
+from apache_beam.utils.timestamp import Duration
+from apache_beam.utils.timestamp import Timestamp
+
+__all__ = [
+    'DeduplictaionWithinDuration',
+]
+
+
+class DeduplictaionWithinDuration(ptransform.PTransform):
+  """ A PTransform which deduplicates input records over a time domain and
+  threshold. Values in different windows will NOT be considered duplicates of
+  each other. Deduplication is best effort.
+
+  The durations specified may impose memory and/or storage requirements within
+  a runner and care might need to be used to ensure that the deduplication time
+  limit is long enough to remove duplicates but short enough to not cause
+  performance problems within a runner. Each runner may provide an optimized
+  implementation of their choice using the deduplication time domain and
+  threshold specified.
+
+  Does not preserve any order the input PCollection might have had.
+  """
+  def __init__(
+      self, time_domain=TimeDomain.REAL_TIME, duration=Duration(10 * 60)):
+    if time_domain not in (TimeDomain.WATERMARK, TimeDomain.REAL_TIME):
+      raise ValueError(
+          'Unsupported TimeDomain: %r. DeduplictaionWithinDuration'
+          'expects TimeDomain.WATERMARK or TimeDomain.REAL_TIME' %
+          (time_domain, ))
+    self.time_domain = time_domain
+    self.duration = duration
+
+  def _create_deduplicate_fn(self):
+    timer_spec = userstate.TimerSpec('expiry_timer', self.time_domain)
+    state_spec = userstate.BagStateSpec('seen', BooleanCoder())
+    duration = self.duration
+    domain = self.time_domain
+
+    class DeduplicationFn(DoFn):
+      def process(
+          self,
+          element,
+          ts=DoFn.TimestampParam,
+          seen_state=DoFn.StateParam(state_spec),
+          expiry_timer=DoFn.TimerParam(timer_spec)):
+        if True in seen_state.read():
+          return
+
+        if domain == TimeDomain.REAL_TIME:
+          expiry_timer.set(Timestamp.now() + duration)
+        elif domain == TimeDomain.WATERMARK:
+          expiry_timer.set(ts + duration)
+        else:
+          raise ValueError(
+              'Unsupported TimeDomain: %r. DeduplictaionWithinDuration'
+              'expects TimeDomain.WATERMARK or TimeDomain.REAL_TIME' %
+              (domain, ))
+        seen_state.add(True)
+        value, _ = element
+        yield value
+
+      def infer_output_type(self, input_type):
+        key_type, _ = trivial_inference.key_value_types(input_type)
+        return key_type
+
+      @userstate.on_timer(timer_spec)
+      def process_timer(self, seen_state=DoFn.StateParam(state_spec)):
+        seen_state.clear()
+
+    return DeduplicationFn()
+
+  def expand(self, pcoll):
+    return (
+        pcoll
+        | 'KeyByElement' >> Map(lambda x: (x, None))
+        | 'DeduplicateFn' >> ParDo(self._create_deduplicate_fn()))
```

Comment:
I would expose a DeduplicatePerKey transform which emits a single KV per key, and then have the generic Deduplicate build on that.

Reply:
Are you suggesting that we provide both `DeduplicatePerKey` and `Deduplicate` which uses value as key?